stackrox/stackrox
# ROX-14694: Add deletion to the reconciliation of declarative configuration

## Description

This PR adds deletion to the reconciliation loop of declarative configurations.

Within the reconciliation, we want to delete each resource that's been created declaratively (i.e. `GetTraits().GetOrigin() == DECLARATIVE`) _and_ that isn't currently provided within the map of transformed messages from the handlers.

During the deletion, if it was successful, the integration health status will be removed for the resource. In case the deletion was unsuccessful (due to any reason), the integration health status is updated and set to unhealty.

Note that currently errors may occur when trying to delete resources that hold stale refernces / would create stale references in other resources (i.e. deleting an access scope right now may not work if it is referenced within a role). For this to work there is [ROX-14700](https://issues.redhat.com/browse/ROX-14700) as a follow-up.

Some additional caveats within this PR:
- Currently, if any error occurs during removing the integration health, this is logged. Since we won't have any other way of notifying this, and this admittedly shouldn't happen, we may think about using `utils.Should` there.
- The reconciliation loop will trigger the deletion for each run, even if the contents of the transformed messages may not have been changed. For this, we might think about either a) creating a hash of the map and only reconciling if there have been changes compared to the previous contents b) use a signal 
that will be reset once reconciliation ran. However, this might be some optimization we could either take within this PR, or at a latter stage, since the stores we are currently working with are relatively low-volume.
- The consecutive error threshold has been lowered, as the current number was too much to handle within tests (also, from a UX point of view, waiting 5 minutes for an error to show up might not be the best choice).

## Checklist

If any of these don't apply, please comment below.

## Testing Performed

- see added unit tests.

Manual tests:
1. Create a central installation with the feature flag turned on and some preliminary config map mounts added:
```bash
roxctl helm output central-services --image-defaults=opensource && helm install stackrox-central-services -n stackrox -f - <<EOF
central:
  declarativeConfiguration:
    mounts:
      configMaps:
        - permission-set
        - access-scope
        - auth-provider
        - role
customize:
  envVars:
    ROX_DECLARATIVE_CONFIGURATION: true
    LOGLEVEL: Debug
EOF ./stackrox-central-services-chart
```
2. Create a config map containing a permission set:
```bash
kubectl apply -f - <<EOF
apiVersion: v1
kind: ConfigMap
metadata:
  name: permission-set
  namespace: stackrox
data:
  permission-set: |
    name: hello-world
    description: the first declaratively created permission set
    resources:
    - resource: Integration
      access: READ_ACCESS
    - resource: Administration
      access: READ_ACCESS
    - resource: Access
      access: READ_ACCESS
EOF
```
3. Verify that the resource is created and shown as healthy:
```bash
roxcurl /v1/integrationhealth/declarativeconfigs | jq
{
  "integrationHealth": [
...
    {
      "id": "89e43e8c-a50b-5429-b53d-13c26b9d5080",
      "name": "hello-world in config map permission-set",
      "type": "DECLARATIVE_CONFIG",
      "status": "HEALTHY",
      "errorMessage": "",
      "lastTimestamp": "2023-03-13T03:18:46.819066029Z"
    }
}
```
4. Delete the config map. Verify that the resource is removed and no health is indicated for it anymore:
```bash
k delete -f - <<EOF
apiVersion: v1
kind: ConfigMap
metadata:
  name: permission-set
  namespace: stackrox
data:
  permission-set: |
    name: hello-world
    description: the first declaratively created permission set
    resources:
    - resource: Integration
      access: READ_ACCESS
    - resource: Administration
      access: READ_ACCESS
    - resource: Access
      access: READ_ACCESS
EOF
roxcurl /v1/integrationhealth/declarativeconfigs | jq
{
  "integrationHealth": [
    {
      "id": "/run/stackrox.io/declarative-configuration/access-scope",
      "name": "Config Map access-scope",
      "type": "DECLARATIVE_CONFIG",
      "status": "UNINITIALIZED",
      "errorMessage": "",
      "lastTimestamp": "2023-03-13T03:56:30.006239172Z"
    },
    {
      "id": "/run/stackrox.io/declarative-configuration/auth-provider",
      "name": "Config Map auth-provider",
      "type": "DECLARATIVE_CONFIG",
      "status": "UNINITIALIZED",
      "errorMessage": "",
      "lastTimestamp": "2023-03-13T03:56:30.058972260Z"
    },
    {
      "id": "/run/stackrox.io/declarative-configuration/role",
      "name": "Config Map role",
      "type": "DECLARATIVE_CONFIG",
      "status": "UNINITIALIZED",
      "errorMessage": "",
      "lastTimestamp": "2023-03-13T03:56:30.163955534Z"
    },
    {
      "id": "/run/stackrox.io/declarative-configuration/permission-set",
      "name": "Config Map permission-set",
      "type": "DECLARATIVE_CONFIG",
      "status": "HEALTHY",
      "errorMessage": "",
      "lastTimestamp": "2023-03-13T04:01:55.162144934Z"
    }
  ]
}
```
5. Re-create the permission set again:
```bash
kubectl apply -f - <<EOF
apiVersion: v1
kind: ConfigMap
metadata:
  name: permission-set
  namespace: stackrox
data:
  permission-set: |
    name: hello-world
    description: the first declaratively created permission set
    resources:
    - resource: Integration
      access: READ_ACCESS
    - resource: Administration
      access: READ_ACCESS
    - resource: Access
      access: READ_ACCESS
EOF
```
7. Verify that the resource is created and shown as healthy:
```bash
roxcurl /v1/integrationhealth/declarativeconfigs | jq
{
  "integrationHealth": [
...
    {
      "id": "89e43e8c-a50b-5429-b53d-13c26b9d5080",
      "name": "hello-world in config map permission-set",
      "type": "DECLARATIVE_CONFIG",
      "status": "HEALTHY",
      "errorMessage": "",
      "lastTimestamp": "2023-03-13T04:04:45.064704234Z"
    },
}
```
8. Create an access scope + role:
```bash
kubectl apply -f - <<EOF
apiVersion: v1
kind: ConfigMap
metadata:
  name: access-scope
  namespace: stackrox
data:
  access-scope: |
    name: hello-world
    description: the first declaratively created access scope
    rules:
      included:
      - cluster: remote
EOF
kubectl apply -f - <<EOF
apiVersion: v1
kind: ConfigMap
metadata:
  name: role
  namespace: stackrox
data:
  role: |
    name: hello-world
    description: the first declaratively created role
    permissionSet: hello-world
    accessScope: hello-world
EOF
```
9. Delete the access scope, this should lead to an unhealthy status for the access scope due to the existing reference in the role:
```bash
roxcurl /v1/integrationhealth/declarativeconfigs | jq
{
  "integrationHealth": [
...
    {
      "id": "df5d408f-e083-5920-b35a-774684481925",
      "name": "hello-world",
      "type": "DECLARATIVE_CONFIG",
      "status": "UNHEALTHY",
      "errorMessage": "cannot delete access scope in use by role \"hello-world\": referenced by another object",
      "lastTimestamp": "2023-03-13T04:07:00.087442675Z"
    }
}
```
